jurity/recommenders/intralist_diversity.py

```diff
@@ -0,0 +1,255 @@
+        it is called on each pair of instances (rows) and the resulting value recorded.
+        The callable should take two arrays from X as input and return a value indicating the distance between them.
+    num_runs: int
+        num_runs is used to report the approximation of Intra-List Diversity over multiple runs on smaller
+        samples of users, default=10, for a speed-up on evaluations. The sampling size is defined by
+        user_sample_size. The final result is averaged over the multiple runs.
+
+    Returns
+    -------
+    Intra-list diversity metric, number of unique users as the support to get the metric
+    """
+
+    # Sample users
+    if user_sample_size is not None:
+        results_over_runs = []
+        supports_over_runs = []
+
+        # Create a different seed for each run
+        rng = np.random.default_rng(seed)
+        seeds = rng.integers(0, num_runs * 10, num_runs)
+
+        for i in range(num_runs):
+            df = sample_users(predicted_results, user_id_column, user_sample_size, seed=seeds[i])
+
+            res, support = intralist_diversity(df, item_features, click_column, k, user_id_column=user_id_column,
+                                               item_id_column=item_id_column, user_sample_size=None,
+                                               metric=metric)
+            results_over_runs.append(res)
+            supports_over_runs.append(support)
+
+        intra_list_diversity = np.mean(results_over_runs)
+        support = int(np.mean(supports_over_runs))
+
+        return intra_list_diversity, support
+
+    if k == 1:
+        warnings.warn('Intra-List Diversity will be nan when only one item is provided in item lists.')
+
+    df = predicted_results
+
+    # Sort by user and score, and take the top k scores.
+    df = get_sorted_clicks(df, user_id_column, click_column, k)
+
+    # mapping user_ids and item_ids to indices
+    unique_user_ids = list(df[user_id_column].unique())
+    user_id_map = dict(zip(unique_user_ids, range(len(unique_user_ids))))
+
+    item_features = item_features.reset_index(drop=True)
+    item_id_map = {v: k for k, v in item_features[item_id_column].to_dict().items()}
+
+    df['item_index'] = df[item_id_column].map(item_id_map).values
+    df['user_index'] = df[user_id_column].map(user_id_map).values
+
+    # cross join on user_id to get all 2-item combinations of items in lists
+    df_merged = df.merge(df, on=user_id_column)
+
+    # remove item pairs with the same items
+    df_merged = df_merged[df_merged['item_index_x'] != df_merged['item_index_y']][
+        [user_id_column, 'item_index_x', 'item_index_y']]
+
+    ## recompute pairwise distances for all combinations of items in the features dataframe
+    item_feature_minus = item_features.set_index(item_id_column)
+    cosine_distance = pairwise_distances(item_feature_minus, metric=metric)
```

Comment:
This function can also take `n_jobs` argument. Given the number of items and feature dimension usually would not be too large to handle, this might not be needed. But if inter and intra get merged together, this is a parameter shared in common.